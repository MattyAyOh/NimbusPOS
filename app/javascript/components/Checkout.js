import React from "react"
import styled from "styled-components"
import { Link } from "react-router-dom"
import moment from "moment"
import jquery from "jquery"

import LineItem from "./LineItem"
import Divider from "./Divider"

class Checkout extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      cash_handled: props.order.cash_handled,
    }
  }

  render() {
    const hours_spent = moment(this.props.order.end_time)
      .diff(this.props.order.start_time, "hours")

    return (
      <Layout>
        <Bill>
          <LineItem
            key={this.props.service.name}
            name={this.props.service.name}
            rate={`${this.props.service.hourly_rate} / hr`}
            quantity={`${hours_spent} hr`}
            amount={this.props.service.hourly_rate * hours_spent}
          />

          {this.props.extras.map((extra) => (
            <LineItem
              key={extra.extra.name}
              name={extra.extra.name}
              rate={extra.extra.price}
              quantity={extra.quantity}
              amount={extra.quantity * extra.extra.price}
            />
          ))}

          <Divider />

          <LineItem
            key="total"
            name="total"
            amount={this.props.order.bill_amount}
          />
        </Bill>

        <div>
          Cash Handled:
          <input
            value={this.state.cash_handled || ""}
            onChange={(event) => this.setState({cash_handled: event.target.value})}
          />
          <Link to="/" onClick={this.persist}>Confirm</Link>
        </div>
      </Layout>
    )
  }

  persist = () => {
    let state = this.state
    let params = this.props.params

    this.setState({ ...state, persisted: false })

    jquery.ajax({
      url: "/update/order",
      type: "PUT",
      data: { params, state },
      success: (response) => this.setState({ persisted: response.persisted }),
    })
  }
}

const Bill = styled.div`
  display: grid;
`

const Layout = styled.div`
  display: grid;
`

export default Checkout
