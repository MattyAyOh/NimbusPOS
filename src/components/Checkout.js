import React from "react"
import PropTypes from "prop-types"
import styled from "styled-components"
import moment from "moment"
import { observable } from "mobx"
import { observer } from "mobx-react"

import Order from "../data/Order"
import LineItem from "./LineItem"

const blue = "#4a90e2"

@observer
class Checkout extends React.Component {
  @observable cash_handled = null

  constructor(props) {
    super(props)

    this.cash_handled = props.order.cash_handled
  }

  render() {
    const hours_spent = (
      moment(this.props.order.end_time)
        .diff(this.props.order.start_time, "minutes") + 1
    ) / 60.0

    return (
      <Layout>
        <Bill>
          <LineItem
            key={this.props.order.service.name}
            name={this.props.order.service.name}
            rate={`${this.props.order.service.hourly_rate} / hr`}
            quantity={`${hours_spent.toFixed(1)} hr`}
            amount={(this.props.order.service.hourly_rate * hours_spent).toFixed(2)}
          />

          {this.props.extras.map((extra) => (
            <LineItem
              key={extra.extra.name}
              name={extra.extra.name}
              rate={extra.extra.price}
              quantity={extra.quantity}
              amount={extra.quantity * extra.extra.price}
            />
          ))}

          <Divider />

          <LineItem
            key="total"
            name="Total"
            amount={this.props.order.bill_amount(this.props.order.service.hourly_rate, moment())}
          />
        </Bill>

        <Register>
          Cash Handled:

          <input
            type="number"
            value={this.cash_handled || ""}
            onChange={(event) => this.set_cash_handled(event.target.value)}
          />
        </Register>

        <Confirm
          onClick={() => this.props.persist({ cash_handled: this.cash_handled })}
          disabled={this.cash_handled == null}
        >Confirm</Confirm>
      </Layout>
    )
  }

  componentDidMount() {
    this.props.onMount()
  }

  set_cash_handled(value) {
    if(value === "")
      this.cash_handled = null
    else
      this.cash_handled = value
  }
}

const Bill = styled.div`
  display: grid;
`

const Layout = styled.div`
  display: grid;
`

const Register = styled.div`
  display: flex;
  justify-content: space-between;
`

const Confirm = styled.button`
  font-size: 1.2rem;
  margin-top: 2rem;
  text-align: center;
`

const Divider = styled.div`
  border-top: 1px solid ${blue};
  height: 1rem;
`

Checkout.propTypes = {
  order: PropTypes.instanceOf(Order),
}

export default Checkout
