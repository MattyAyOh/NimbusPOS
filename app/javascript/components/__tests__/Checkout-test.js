import Adapter from 'enzyme-adapter-react-16';
import Enzyme from 'enzyme';
import React from "react";
import renderer from 'react-test-renderer';

import Checkout from "../Checkout";

Enzyme.configure({ adapter: new Adapter() });

// Snapshot test - https://facebook.github.io/jest/docs/en/snapshot-testing.html
test("Displays correctly", () => {
  const tree = renderer.create(<Checkout {...default_props} />).toJSON();
  expect(tree).toMatchSnapshot();
})

test("Confirm button is disabled until 'Cash Handled' is inputted", () => {
  const checkout = Enzyme.shallow(<Checkout {...default_props} />);
});

const default_props = {
  onMount: () => {},
  order: {
    extras: [],
    cash_handled: null,
    end_time: "2018-02-24T21:00:00.000Z",
    start_time: "2018-02-24T20:00:00.000Z",
  },
  service: {
    hourly_rate: 20,
    name: "Mahjong",
    position: 1,
    service: "mahjong",
  },
  extras: [],
}
