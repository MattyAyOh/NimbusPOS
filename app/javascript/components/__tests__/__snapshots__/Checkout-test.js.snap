// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Displays correctly 1`] = `
<div
  className="sc-bxivhb etFryt"
>
  <div
    className="sc-htpNat ebTkst"
  >
    <div
      className="sc-bdVaJa jzqtRL"
    >
      <span>
        Mahjong
      </span>
      <span>
        $20 / hr
      </span>
      <span>
        1 hr
      </span>
      <span
        className="sc-bwzfXH dkRrwr"
      >
        $
        20
      </span>
    </div>
    <div
      className="sc-bZQynM jAZOmj"
    />
    <div
      className="sc-bdVaJa jzqtRL"
    >
      <span>
        Total
      </span>
      <span />
      <span />
      <span
        className="sc-bwzfXH dkRrwr"
      >
        $
        20.00
      </span>
    </div>
  </div>
  <div
    className="sc-ifAKCX fOGIFb"
  >
    Cash Handled:
    <input
      onChange={[Function]}
      value=""
    />
  </div>
  <span
    className="sc-EHOje jjUlfT"
    onClick={[Function]}
  >
    Confirm
  </span>
</div>
`;
